### ISSUE TITLE
Solution jupyter vs .py

### ISSUE THREAD

**Issue by georgebuzzell** [NDCLab/baseEEG#21]
Solution to allow development using either .py or jupyter

**Comment by Jonhas** [comment]
@georgebuzzell There is a package known as ipynb-py-convert, which allows to convert .py files to jupyter notebook files. We can install this package by doing the following
`pip3 install ipynb-py-convert`
we then want to add this to our path, so we can be able to run the script without executing ~/.local/bin/./ipynb-py-convert source.py dest.py. This wall also allow us to run any of our packages as simple scripts without specifying the complete path. We do the following 
`export PATH=$PATH:/home/ubuntu/.local/bin` 
`source .bashrc`
We make sure to run source ~/.bashrc to preserve the modifications to our current shell. If running zsh, apply changes to your ~/.zshrc 
Now if we were to run ipynb-py-convert alone, we should receive the following message: 
`Usage: ipynb-py-convert in.ipynb out.py
or:    ipynb-py-convert in.py out.ipynb`
It's a little annoying to write out the entire script name, so we can make an alias for the script. we add the following to the end 
of our bashrc. 
`alias pynb [or any name] = 'ipynb-py-convert' `
`source ~/.bashrc `
Consider we have the following script
```python
#!/usr/bin/python3 
import os
import mne

sample_data_folder = mne.datasets.sample.data_path()
sample_data_raw_file = os.path.join(sample_data_folder, 'MEG', 'sample',
                                    'sample_audvis_raw.fif')
raw = mne.io.read_raw_fif(sample_data_raw_file)
raw.plot()
```
When we run ./test3.py, we should get the following output 
``` 
Collecting mypy
  Downloading mypy-0.812-cp38-cp38-manylinux2010_x86_64.whl (22.5 MB)
     |████████████████████████████████| 22.5 MB 22.2 MB/s 
Collecting typed-ast<1.5.0,>=1.4.0
  Downloading typed_ast-1.4.3-cp38-cp38-manylinux1_x86_64.whl (774 kB)
     |████████████████████████████████| 774 kB 12.1 MB/s 
Collecting mypy-extensions<0.5.0,>=0.4.3
  Downloading mypy_extensions-0.4.3-py2.py3-none-any.whl (4.5 kB)
Requirement already satisfied, skipping upgrade: typing-extensions>=3.7.4 in /usr/local/lib/python3.8/dist-packages (from mypy) (3.7.4.3)
Installing collected packages: typed-ast, mypy-extensions, mypy
  WARNING: The scripts dmypy, mypy, stubgen and stubtest are installed in '/home/ubuntu/.local/bin' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Successfully installed mypy-0.812 mypy-extensions-0.4.3 typed-ast-1.4.3
ubuntu@b8c3b29b605d:~/test$ ls
test3.ipynb  test3.py
ubuntu@b8c3b29b605d:~/test$ ./test3.py 
Using default location ~/mne_data for sample...
Creating ~/mne_data
Downloading archive MNE-sample-data-processed.tar.gz to /home/ubuntu/mne_data
Downloading https://files.osf.io/v1/resources/rxvq7/providers/osfstorage/59c0e26f9ad5a1025c4ab159?version=5&action=download&direct (1.54 GB)
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| Downloading : 1.54G/1.54G [00:54<00:00,    30.2MB/s]
Verifying hash 12b75d1cb7df9dfb4ad73ed82f61094f.
Decompressing the archive: /home/ubuntu/mne_data/MNE-sample-data-processed.tar.gz
(please be patient, this can take some time)
Successfully extracted to: ['/home/ubuntu/mne_data/MNE-sample-data']
Attempting to create new mne-python configuration file:
/home/ubuntu/.mne/mne-python.json
Opening raw data file /home/ubuntu/mne_data/MNE-sample-data/MEG/sample/sample_audvis_raw.fif...
    Read a total of 3 projection items:
        PCA-v1 (1 x 102)  idle
        PCA-v2 (1 x 102)  idle
        PCA-v3 (1 x 102)  idle
    Range : 25800 ... 192599 =     42.956 ...   320.670 secs
```
Now, we can run the following: 
```bash  
pynb test3.py test3.ipynb 
```
which should produce a separate jupyter notebook file. Consider the following figure. 
![image](https://user-images.githubusercontent.com/45021859/114500131-182ba500-9bed-11eb-90a9-3106082f9b8b.png)
![image](https://user-images.githubusercontent.com/45021859/114500199-32fe1980-9bed-11eb-8a8a-403b6cd0a180.png)
However, An issue though is that it all gets translated to one cell only.
How this can work is that we can have a script run after every merge to the dev branch. We can do the following
```vim
touch ~/[baseEEG project name]/.git/hooks/post-merge && chmod +x ~/[baseEEG project name]/.git/hooks/post-merge
```
I'm currently working on the script and would need to do some testing.

**Comment by georgebuzzell** [comment]
@Jonhas This looks promising and I really appreciate you working on this! That said, I think I would need to chat with you to more completely understand. Is it worth us jumping on a quick call to touch base about this? Or do you want to complete the additional testing first? Or, would this all be best to discuss at the upcoming group meeting? Please let me know, one way or another. Either way seems like you are onto a creative solution here. :)

**Comment by Jonhas** [comment]
@georgebuzzell I would like to finish some additional testing before I present it since I was able to find a better package to use, p2j, which parses a python script and converts it to a jupyter notebook. How p2j differs from ipynb-py-convert is that each block of code turns into a cell. This is already better than the previous package, which simply converts the entire script to one cell. So I wanted to make those respective changes. Additionally, since it is possible that someone will push more than one *.py file or only a *.ipynb file, I am currently working on the post-merge hook as mentioned prior to convert some list of git diffs of *.py files to convert to their respective *.ipynb file.  Furthermore, if only a *.ipynb file was merged, we would want to convert it to its respective *.py file.
